File: api/algorithms/ucs.py
import heapq

DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def ucs(start, goal, obstacles, rows, cols):
    pq = [(0, start, [])]
    visited = set()

    while pq:
        cost, current, path = heapq.heappop(pq)
        if current in visited:
            continue
        visited.add(current)
        if current == goal:
            return path
        for dx, dy in DIRECTIONS:
            pos = (current[0] + dx, current[1] + dy)
            if 0 <= pos[0] < rows and 0 <= pos[1] < cols and pos not in obstacles and pos not in visited:
                heapq.heappush(pq, (cost + 1, pos, path + [(dx, dy)]))
    return []

File: client/src/algorithms/ucs.js
const DIRS = [[-1, 0], [1, 0], [0, -1], [0, 1]];

export function ucs(start, goal, obstacles, rows, cols) {
  let pq = [[0, start, []]];
  const visited = new Set();

  while (pq.length > 0) {
    pq.sort((a, b) => a[0] - b[0]);
    const [cost, cur, path] = pq.shift();
    const key = `${cur[0]},${cur[1]}`;
    if (visited.has(key)) continue;
    visited.add(key);
    if (cur[0] === goal[0] && cur[1] === goal[1]) return path;

    for (const [dr, dc] of DIRS) {
      const nr = cur[0] + dr, nc = cur[1] + dc;
      const nk = `${nr},${nc}`;
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && !obstacles.has(nk) && !visited.has(nk)) {
        pq.push([cost + 1, [nr, nc], [...path, [dr, dc]]]);
      }
    }
  }
  return [];
}
